packages/react-reconciler/src/ReactFiberCommitWork.js

```diff
@@ -3774,6 +3801,32 @@ function commitOffscreenPassiveMountEffects(
   }
 }
 
+function commitFragmentInstanceInsertionEffects(finishedWork: Fiber): void {
+  if (enableFragmentRefs && finishedWork.tag === HostComponent) {
+    const parentFragmentInstances = getFragmentInstanceParents(finishedWork);
```

Comment:
This will allocate an array every time you add/remove something in a fragment which can be many many times a second. Thrashing the GC more. You can avoid this allocation by using the stack. E.g. just inline this and do the work in the loop as you want the return path.